TTDQueries/tutorial-instructions.md

```diff
@@ -0,0 +1,243 @@
+
+Can we find out the exact point in time where the program created the message box with
+the confusing error?
+
+One way to think about this problem is to list all calls to `user32!MessageBoxW` (Windows
+API to show a message box), order it by the start time of the function, and then pick the
+last call. That is the point in time where we'll be able to inspect where the error came from:
+
+```LINQ
+dx @$cursession.TTD.Calls("user32!MessageBoxW").OrderBy(c => c.TimeStart).Last()
+
+Result:
+@$cursession.TTD.Calls("user32!MessageBoxW").OrderBy(c => c.TimeStart).Last()
+    EventType        : Call
+    ThreadId         : 0x6c4
+    UniqueThreadId   : 0x2
+    TimeStart        : 306600:5BF [Time Travel]
+    TimeEnd          : 30CADE:80 [Time Travel]
+    Function         : UnknownOrMissingSymbols
+    FunctionAddress  : 0x7ffb578f1e30
+    ReturnAddress    : 0x14000c8d1
+    ReturnValue      : 0x1
+    Parameters
+```
+
+We can navigate to that position in time in order to figure out where the error came from.
+You can click on the time link ('Time Travel').
+
+Look at the stack and change your frame to frame 0x1, the caller of `user32!MessageBoxW`, by
+double clicking on it.
+
+<img src="images/windbg-message-box-call.jpg" alt="Windbg Preview Message Box Call" width="1200"/>
+
+As you can see the error comes from calling GetLastError(). At the point when
+GetLastError() is called the real error has been reset. This is likely to happen
+if you don't call GetLastError() right after the Windows API that reports the
+issue and store the result in a variable.
+
+The next frame down in the stack is our program trying to open a file. We can step back from
+this point, to figure out what file it was trying to open.
+To do this, click on the backwards 'step over' control in the ribbon of Windbg Preview.
+
+<img src="images/windbg-ttd-controls.jpg" alt="Windbg Preview TTD Controls" width="900"/>
+
+Now that we have stepped backwards, we are able to see what file `DShowPlayer!MainWindow::OnFileOpen()`
+was trying to open.
+
+<img src="images/windbg-invalid-file.jpg" alt="Windbg Preview Invalid File" width="1200"/>
+
+### Using queries to discover patterns in execution data
+
+Maybe you would like to understand what is the pattern arround the error's data.
+You can write queries and display them in a grid for better visualization. The `-g`
+option in dx allows you to enumerable data in a grid display.
+
+We could a get general sense for the errors encountered in our program by looking
+at the count of each error reported by `GetLastError()`.
+We do this by looking for all the calls to the function `KernelBase::GetLastError()`,
+removing calls that return success (value of 0), groping the calls by their return,
+and lastly ordering them by frequency.
+
+* Note that queries can become complex, but you always start simple and build on
+it as determine how you want to slice the data available in the recording.
+
+```LINQ

```

Comment:
It might make sense to walk through the creation of this query more incrementally.

Reply:
Ok. I'll do that walk through in a separate PR. I want to make sure most of the content is in.